src/shared/discover-image.c

```diff
@@ -818,47 +828,42 @@ int image_find(RuntimeScope scope,
+                                        log_debug_errno(r, "Failed to pick versioned image on '%s/%s', skipping: %m", strempty(root), skip_leading_slash(fname_path));
                                         continue;
                                 }
                                 if (!result.path) {
-                                        log_debug("Found versioned directory '%s', without matching entry, skipping.", vp);
+                                        log_debug("Found versioned directory '%s/%s', without matching entry, skipping.", strempty(root), skip_leading_slash(fname_path));
                                         continue;
                                 }
 
                                 /* Refresh the stat data for the discovered target */
                                 st = result.st;
                                 close_and_replace(fd, result.fd);
-
-                                _cleanup_free_ char *bn = NULL;
-                                r = path_extract_filename(result.path, &bn);
-                                if (r < 0) {
-                                        log_debug_errno(r, "Failed to extract basename of image path '%s', skipping: %m", result.path);
-                                        continue;
-                                }
-
-                                fname_buf = path_join(fname, bn);
-                                if (!fname_buf)
+                                resolved_file = mfree(resolved_file);
+                                resolved_file = path_join(root, result.path);
+                                if (!resolved_file)
                                         return log_oom();
 
-                                fname = fname_buf;
-
+                                /* fname and fname_path are invalid now because they
+                                 * would need to be set from result.path by extracting
+                                 * the filename to set fname = path_join(fname, filename)
+                                 * and then fname_path = path_join(*s, fname) but
+                                 * since they are unused we don't do it */
                         } else if (!S_ISDIR(st.st_mode) && !S_ISBLK(st.st_mode)) {
                                 log_debug("Ignoring non-directory and non-block device file '%s' without suffix.", fname);
                                 continue;
                         }
 
-                        _cleanup_free_ char *path = path_join(resolved, fname);
-                        if (!path)
-                                return -ENOMEM;
-
-                        r = image_make(class, name, fd, path, &st, ret);
+                        /* Only put resolved paths into the image entry (incl. --root=).
+                         * Defending against symlink races is out of scope
+                         * and we trust a given root in that regard. */
```

Comment:
I wouldn't say we trust a given root in that regard. I'd rather say this is just TODO to implement.